blogposts/2021/tackling-the-long-tail.md

```diff
@@ -0,0 +1,104 @@
+
+In this post we give you a brief overview of how we're tackling the long tail of tiny repositories by introducing shard
+merging to Zoekt, in
+our [quest to index the OSS universe](https://about.sourcegraph.com/blog/why-index-the-oss-universe/).
+
+## What is Zoekt?
+
+[Zoekt](https://github.com/sourcegraph/zoekt) is a code search engine that
+performs trigram-based regex search. Originally created by Han-Wen Niehuys, Zoekt is fast, easy to deploy, and easy to maintain, which makes it a great choice for
+our self-hosted customer deployments. [Sourcegraph is actually taking on maintainership of Zoekt, which you can read about here](ADD LINK).
+
+## Why not just scale out?
+
+Naturally, Zoekt's size of the index scales with the size of the input data. Zoekt’s index is about 2–3 times larger
+than the input data. Some of that data, like the trigrams, is kept in memory. At the scale of the open source universe,
+it quickly becomes too costly to just scale out. Luckily, Zoekt still has a lot of untapped potential when it comes to
+more efficient data structures. For example, in
+a previous post, Ryan Hitchman
+explained [how we changed one of Zoekt's core data structures to reduce memory by 5x](https://about.sourcegraph.com/blog/zoekt-memory-optimizations-for-sourcegraph-cloud/).
+
+The core idea of this project is to merge the long tail of small, stale repositories into more efficient representations
+on disk and in memory. To understand the motivation behind this idea we have to first dive into Zoekt's data model.
+
+## Zoekt’s data model
+
+Zoekt indexes repositories and stores the index in one or more files on disk. A repository contains a hierarchy of
+documents, such as source code, binaries, text files and so forth. At
+index time [documents are added, one at a time, to the index builder](https://sourcegraph.com/github.com/sourcegraph/zoekt@6a4adda25a6c5a7c6612e309249420102c587b4d/-/blob/gitindex/index.go?L498-505). Once we cross a threshold of input data, currently configured to be 100 MiB, the
+builder [flushes the index to a file on disk](https://sourcegraph.com/github.com/sourcegraph/zoekt@6a4adda25a6c5a7c6612e309249420102c587b4d/-/blob/build/builder.go?L455-457). This file is called a shard. For small repositories, there is a 1:1 relationship between the repository and its shard.
+Larger repositories, such as [Kubernetes](https://sourcegraph.com/github.com/kubernetes/kubernetes) or the [Linux Kernel](https://sourcegraph.com/github.com/torvalds/linux) map to several shards. At query time, shards are treated independently.
+
+## The long tail
+
+Most repositories are tiny and fit within a single shard. The following histogram shows the distribution of shard sizes
+on one of our production instances.
+
+<figure>
+  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/tackling-long-tail/tackling-long-tail-histogram.png" alt="Distribution of shard sizes" class="no-shadow">
+</figure>
```

Comment:
@stefanhengl Just an FYI, I changed how the images are included so we can add captions and remove the drop shadow, which always looks a bit odd to me on white-background images!